File: files/300-399/393/CH-393.R
library(tidyverse)
library(readxl)

path <- "300-399/393/CH-393 Table Transformation.xlsx"
input <- read_excel(path, range = "B3:E8")
test <- read_excel(path, range = "G3:J9")

result = input %>%
  t() %>%
  as.data.frame() %>%
  janitor::row_to_names(row_number = 1) %>%
  pivot_longer(
    cols = -c(Customer, Product),
    names_to = "Date",
    values_to = "Sale"
  ) %>%
  na.omit() %>%
  arrange(Date) %>%
  mutate(Sale = as.numeric(Sale)) %>%
  select(Date, Customer, Product, Sale)

all.equal(result, test, check.attributes = FALSE)
# [1] TRUE

File: files/300-399/393/CH-393.py
import pandas as pd

path = "300-399/393/CH-393 Table Transformation.xlsx"
input = pd.read_excel(path, usecols="B:E", nrows=6, skiprows=2)
test = pd.read_excel(path, usecols="G:J", nrows=7, skiprows=2)


def transform(input_df: pd.DataFrame) -> pd.DataFrame:
    df = input_df.T.copy()
    df.columns = df.iloc[0]
    df = df.iloc[1:].reset_index(drop=True)

    result = (
        df.melt(id_vars=["Customer", "Product"], var_name="Date", value_name="Sale")
          .dropna(subset=["Sale"])
          .sort_values("Date")
          .assign(Sale=lambda d: pd.to_numeric(d["Sale"], errors="coerce"))
          .loc[:, ["Date", "Customer", "Product", "Sale"]]
          .reset_index(drop=True)
    )
    return result

result = transform(input)
print(result.equals(test))
# True